# __Working on Forward Propagation__
In the context of neural networks, forward propagation refers to the process of passing input data through the network's layers one by one to compute the corresponding output. During forward propagation, each layer performs a series of mathematical operations, such as matrix multiplications and activation functions, to transform the input and generate an output. This process allows the network to make predictions or perform computations based on the given input data.

Let us understand how forward propagation works.

## Steps to Be Followed:
1. Importing the required libraries
2. Initializing the data and predicting the output
3. Calculating the errors
4. Calculating the sum of squared error

### Step 1: Importing the Required Libraries
- Import NumPy package
- Import the Python package matplotlib, which sets the padding between and around the subplots as well as the figure size

In [ ]:
import numpy as np
import matplotlib.pyplot as plt

### Step 2: Initializing the Data and Predicting the Output
- Initializing the data:
  - Generate an array **x** with 11 random integer elements ranging from -1 to 4 using NumPy's **random.randint()** function
- Predicting the output:

  - Multiply each element of **x** by 3 and add 2 to obtain the predicted output **y_pred**
  - Print the predicted values **y_pred**


In [ ]:
x = np.random.randint(-1,5,11)
y_pred = 3*x + 2
print(y_pred)


[-1  8 11  8  8  2 14 -1  8  5  8]


**Observation**
- This shows the predicted y to be -1, 8, 11, 8, 8, 2, 14, -1, 8, 5, 8.


### Step 3: Calculating the Errors
- Set the target value **y** to 1
- Calculate the absolute difference between the predicted values **y_pred** and the target value **y** to obtain the error. Print the error array and compute the sum of errors using the **sum()** function


In [ ]:
y = 1
error = abs(y_pred-y)
print(error)
error.sum()


[ 2  7 10  7  7  1 13  2  7  4  7]


67

**Observation**
- The output represents the array of absolute differences between the predicted values **y_pred** and the target value **y** for each element in the array. The values in the array indicate the absolute errors for each corresponding element.

- The sum of the errors, which is 67, represents the total accumulated error when summing up all the individual absolute differences between the predicted and target values.

### Step 4: Calculating the Sum of Squared Error
- It calculates the sum of squared errors (SSE) by squaring the absolute differences between the predicted values **y_pred** and the target value **y** for each element in the array.

The output **SSE** is an array that contains the squared errors for each corresponding element.

The sum of the squared errors, obtained by calling **SSE.sum()**, represents the total accumulated squared error when summing up all the individual squared differences between the predicted and target values.


In [ ]:
SSE = abs(y_pred-y)**2
print(SSE)
SSE.sum()


[  4  49 100  49  49   1 169   4  49  16  49]


539

**Observation**
- The output of 539 represents the sum of squared errors (SSE) calculated from the squared differences between the predicted values and the target value for each element in the array.